staging/src/k8s.io/apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,194 @@
+	UpdatesPending(group, resource string) bool
+
+	// UpdateStorageVersions updates the StorageVersions.
+	UpdateStorageVersions(loopbackClientConfig *rest.Config, apiserverID string)
+	// Completed returns if updating StorageVersions has completed.
+	Completed() bool
+}
+
+var _ Manager = &DefaultManager{}
+
+// NewDefaultManager creates a new DefaultManager.
+func NewDefaultManager() *DefaultManager {
+	s := &DefaultManager{}
+	s.completed.Store(false)
+	s.groupResources = make(map[string]map[string]struct{})
+	s.resources = make(map[*ResourceInfo]struct{})
+	return s
+}
+
+// AddResourceInfo adds ResourceInfo to the manager.
+// This is not thread-safe. It is expected to be called when the apiserver is installing the endpoints, which is done serially.
+func (s *DefaultManager) AddResourceInfo(resources ...*ResourceInfo) {
+	for _, r := range resources {
+		s.resources[r] = struct{}{}
+		s.addGroupResourceFor(r)
+	}
+}
+
+func (s *DefaultManager) addGroupResourceFor(r *ResourceInfo) {
+	gvrs := r.EquivalentResourceMapper.EquivalentResourcesFor(schema.GroupVersionResource{
+		Group:    r.Group,
+		Resource: r.Resource.Name,
+	}, "")
+	for _, gvr := range gvrs {
+		s.addGroupResource(gvr.Group, gvr.Resource)
+	}
+}
+
+func (s *DefaultManager) addGroupResource(group, resource string) {
+	if _, ok := s.groupResources[group]; !ok {
+		s.groupResources[group] = make(map[string]struct{})
+	}
+	s.groupResources[group][resource] = struct{}{}
+}
+
+// RemoveResourceInfo removes ResourceInfo from the manager.
+// It is not safe to call this function concurrently with AddResourceInfo.
+func (s *DefaultManager) RemoveResourceInfo(r *ResourceInfo) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	delete(s.resources, r)
+	s.removeGroupResourceFor(r)
+}
+
+func (s *DefaultManager) removeGroupResourceFor(r *ResourceInfo) {
+	gvrs := r.EquivalentResourceMapper.EquivalentResourcesFor(schema.GroupVersionResource{
+		Group:    r.Group,
+		Resource: r.Resource.Name,
+	}, "")
+	for _, gvr := range gvrs {
+		s.removeGroupResource(gvr.Group, gvr.Version)
+	}
+}
+
+func (s *DefaultManager) removeGroupResource(group, resource string) {
+	if _, ok := s.groupResources[group]; !ok {
+		return
+	}
+	delete(s.groupResources[group], resource)
+	if len(s.groupResources[group]) == 0 {
+		delete(s.groupResources, group)
+	}
+}
+
+// UpdatesPending returns if the StorageVersion of a resource is still wait to be updated.
+func (s *DefaultManager) UpdatesPending(group, resource string) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if _, ok := s.groupResources[group]; !ok {
+		return false
+	}
+	_, ok := s.groupResources[group][resource]
+	return ok
+}
+
+// DefaultManager indicates if the aggregator, kube-apiserver, and the
+// apiextensions-apiserver have completed reporting their storage versions.
+type DefaultManager struct {
+	completed atomic.Value
+
+	mu             sync.RWMutex
+	resources      map[*ResourceInfo]struct{}
+	groupResources map[string]map[string]struct{}
```

Comment:
How about `groupResources map[schema.GroupResource]resourceStatus`

```go
type resourceStatus struct{
    done bool
    lastErr error
}
```

for our future inspection.

Reply:
This would also eliminate the parallel maps